service/Core/Handlers/EmbeddingGeneratorHelper.cs

```diff
@@ -0,0 +1,154 @@
+        }
+
+        return helper;
+    }
+
+    public EmbeddingGeneratorHelper(List<ITextEmbeddingGenerator> embeddingGenerators)
+    {
+        this._embeddingGenerators = embeddingGenerators ?? throw new System.ArgumentNullException(nameof(embeddingGenerators));
+        this._cache = new();
+        foreach (var embeddingGenerator in embeddingGenerators)
+        {
+            this._cache[embeddingGenerator] = new();
+        }
+    }
+
+    public async Task GenerateEmbeddingAsync(IList<string> text, CancellationToken cancellationToken)
+    {
+        //we will proceed for each generator
+        //we will first check if we can use the batch generation
+        foreach (var generator in this._embeddingGenerators)
+        {
+            //ok now we can calculate the embedding in batch if we have more than one text and batch generator is active.
+            if (text.Count > 1 && generator is ITextEmbeddingBatchGenerator batchGenerator)
+            {
+                List<string> batch = new();
+                int tokenCount = 0;
+                for (int i = 0; i < text.Count; i++)
+                {
+                    var singlePieceOfText = text[i];
+                    var pieceTokenCount = generator.CountTokens(singlePieceOfText);
+
+                    // The next piece of element will exceed the limit, or we reached maximum batch size.
+                    if (BatchReachedMaximumNumberOfElements(generator, batchGenerator, batch, tokenCount, pieceTokenCount))
+                    {
+                        await this.BatchConvertAsync(generator, batchGenerator, batch, cancellationToken).ConfigureAwait(false);
+                        tokenCount = 0; //a new token count starts.
+                    }
+
+                    batch.Add(singlePieceOfText);
+                    tokenCount += pieceTokenCount;
+                }
+
+                //flush the last batch.
+                if (batch.Count > 0)
+                {
+                    await this.BatchConvertAsync(generator, batchGenerator, batch, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            else
+            {
+                //sequentially generate the embeddings, no way to optimize.
+                foreach (var singlePieceOfText in text)
+                {
+                    Embedding embedding = await generator.GenerateEmbeddingAsync(singlePieceOfText, cancellationToken).ConfigureAwait(false);
+                    this._cache[generator][singlePieceOfText] = embedding;
+                }
+            }
+        }
+    }
+
+    private static bool BatchReachedMaximumNumberOfElements(
+        ITextEmbeddingGenerator generator,
+        ITextEmbeddingBatchGenerator batchGenerator,
+        List<string> batch,
+        int tokenCount,
+        int pieceTokenCount)
+    {
+        return tokenCount + pieceTokenCount > generator.MaxTokens || batch.Count == batchGenerator.MaxBatchSize;
```

Comment:
generator.MaxTokens (8191 with ada2) applies to individual strings only, not the batch size. 
For example, in one of my tests I sent a batch of 96 strings for a total of 83380 tokens.